### PR TITLE
Search controller refactoring

### PR DESCRIPTION
**Pull request by vpereira**
Refactor method `SearchController#filter_items` to private and change how to handle `limit` and `offset`. Add spec for this case

-----------------

If this PR requires any particular action or consideration before deployment,
please check the reasons or add your own to the list:

* [ ] Requires a database migration that can cause downtime. Postpone deployment until maintenance window[1].
* [ ] Contains a data migration that can cause temporary inconsistency, so should be run at a specific point of time.
* [ ] Changes some configuration files (e.g. options.yml), so the changes have to be applied manually in the reference server.
* [ ] A new Feature Toggle[2] should be enabled in the reference server.
* [ ] Proper documentation or announcement has to be published upfront since the introduced changes can confuse the users.

[1] https://github.com/openSUSE/open-build-service/wiki/Deployment-of-build.opensuse.org#when-there-are-migrations
[2] https://github.com/openSUSE/open-build-service/wiki/Feature-Toggles-%28Flipper%29#you-want-real-people-to-test-your-feature

### CODE REVIEW

**Comment by vpereira** [comment]
## no offset, no limit

```
$ dosc api '/search/package?match=%40name+%3D+%27Foo%27+or+title+%3D+%27Foo%27+or+description+%3D+%27Foo%27'
WARNING: SSL certificate checks disabled. Connection is insecure!

<collection matches="3">
<package name="Foo" project="home:Admin">
  <title/>
  <description/>
</package>
<package name="Foo" project="Bar">
  <title/>
  <description/>
</package>
<package name="Foo" project="Test">
  <title/>
  <description/>
</package>
</collection>
```
## offset 1 = skip first

```
$ dosc api '/search/package?match=%40name+%3D+%27Foo%27+or+title+%3D+%27Foo%27+or+description+%3D+%27Foo%27&offset=1'
WARNING: SSL certificate checks disabled. Connection is insecure!

<collection matches="3">
<package name="Foo" project="Bar">
  <title/>
  <description/>
</package>
<package name="Foo" project="Test">
  <title/>
  <description/>
</package>
</collection>
```
## offset = 1, limit = 1 (just show the second record)

```
dosc api '/search/package?match=%40name+%3D+%27Foo%27+or+title+%3D+%27Foo%27+or+description+%3D+%27Foo%27&offset=1&limit=1'
WARNING: SSL certificate checks disabled. Connection is insecure!

<collection matches="3">
<package name="Foo" project="Bar">
  <title/>
  <description/>
</package>
</collection>
```

Note that the `collection matches` are always wrong if there is some filtering. That's because we compute it before applying the filter https://github.com/openSUSE/open-build-service/blob/master/src/api/app/controllers/search_controller.rb#L200 is that a bug or feature @adrianschroeter ?

File: src/api/spec/controllers/search_controller_spec.rb [hunk 1]
```diff
@@ -5,19 +5,20 @@
   render_views
 
   let!(:user) { create(:confirmed_user, :with_home, login: 'foo') }
```
Comment (hennevogel) [inline src/api/spec/controllers/search_controller_spec.rb:7]: Now you really don't need to use `let!` anymore..

File: src/api/spec/controllers/search_controller_spec.rb [hunk 2]
```diff
@@ -38,6 +39,21 @@
     subject! { get :package, params: { match: "@name='apacheX'" } }
 
     it_behaves_like 'find package'
+
+    context 'With offset' do
+      subject! { get :package, params: { offset: 'XXX', match: "@name='apacheX'" } }
+
+      it_behaves_like 'find package'
```
Comment (hennevogel) [inline src/api/spec/controllers/search_controller_spec.rb:46]: I'm not sure what you are testing here? That `offset: 'XXX'` has no effect?

File: src/api/spec/controllers/search_controller_spec.rb [hunk 3]
```diff
@@ -38,6 +39,21 @@
...
+    end
+  end
+
+  describe 'search for packages with offset = 1' do
+    let!(:package) { create(:package, name: 'apacheX', project: user.home_project) }
+    let!(:project) { create(:project_with_package, name: 'Foo', maintainer: user, package_name: 'apacheX') }
+
+    subject! { get :package, params: { limit: '1', offset: '1', match: "@name='apacheX'" } }
+
+    it_behaves_like 'find package'
```
Comment (hennevogel) [inline src/api/spec/controllers/search_controller_spec.rb:56]: What about the other cases? Like no `limit` or no `offset`?

**Review by hennevogel — CHANGES_REQUESTED** [review]

File: src/api/spec/controllers/search_controller_spec.rb (hunk 2 above)
Comment (vpereira) [reply src/api/spec/controllers/search_controller_spec.rb:46]: yes, well i added this spec before changing the code.. and before the original code was doing some manual convertion: https://github.com/openSUSE/open-build-service/blob/master/src/api/app/controllers/search_controller.rb#L125, i can drop it now for sure.

File: src/api/spec/controllers/search_controller_spec.rb (hunk 1 above)
Comment (vpereira) [reply src/api/spec/controllers/search_controller_spec.rb]: done

File: src/api/spec/controllers/search_controller_spec.rb (hunk 3 above)
Comment (vpereira) [reply src/api/spec/controllers/search_controller_spec.rb:56]: it's being tested here https://github.com/openSUSE/open-build-service/blob/4dd82b05aad4342e1538d438ea56d3cd74b03328/src/api/spec/controllers/search_controller_spec.rb#L41

File: src/api/spec/controllers/search_controller_spec.rb (hunk 2 above)
Comment (hennevogel) [reply src/api/spec/controllers/search_controller_spec.rb:46]: Yeah a simple `.to_i` should suffice...

File: src/api/spec/controllers/search_controller_spec.rb (hunk 3 above)
Comment (hennevogel) [reply src/api/spec/controllers/search_controller_spec.rb:56]: Hm maybe I don't understand those parameters correctly. `limit=2` would mean out of 5 results I get 2. And `offset=4` would mean I get all results from result 5 onward right? Those cases I mean...

Comment (vpereira) [reply src/api/spec/controllers/search_controller_spec.rb:56]: you can see https://github.com/openSUSE/open-build-service/pull/11397#issuecomment-883320571 where i explain it (or at least try to) with examples... `offset=1` means we do like `results[1..]`. With `limit` you limit the number of results. So `offset=1` and `limit=4`, would mean something like `results[1..].take(4)` 

i.e:

```
2.6.5 :008 > offset=1
 => 1 
2.6.5 :009 > limit=4
 => 4 
2.6.5 :010 > 0.upto(10).to_a[offset..].take(limit)
 => [1, 2, 3, 4] 
2.6.5 :011 > 
```

File: src/api/spec/controllers/search_controller_spec.rb (hunk 2 above)
Comment (vpereira) [reply src/api/spec/controllers/search_controller_spec.rb]: done

File: src/api/spec/controllers/search_controller_spec.rb (hunk 3 above)
Comment (hennevogel) [reply src/api/spec/controllers/search_controller_spec.rb:50]: Okay, so where is this tested?

Comment (hennevogel) [reply src/api/spec/controllers/search_controller_spec.rb:50]: We talked offline, this is in minitest and will be moved later...

**Review by hennevogel — APPROVED** [review]

**Pull request merged**